warehouse/http.go

```diff
@@ -0,0 +1,404 @@
+	pendingUploadCount, err := a.uploadRepo.Count(r.Context(), filters...)
+	if err != nil {
+		a.logger.Errorf("getting pending uploads count", "error", err)
+		http.Error(w, fmt.Sprintf("getting pending uploads count: %s", err.Error()), http.StatusInternalServerError)
+		return
+	}
+
+	filters = []repo.FilterBy{
+		{Key: "source_id", Value: sourceID},
+		{Key: "metadata->>'source_task_run_id'", Value: pendingEventsReq.TaskRunID},
+		{Key: "status", Value: "aborted"},
+	}
+	abortedUploadCount, err := a.uploadRepo.Count(r.Context(), filters...)
+	if err != nil {
+		a.logger.Errorf("getting aborted uploads count", "error", err.Error())
+		http.Error(w, fmt.Sprintf("getting aborted uploads count: %s", err), http.StatusInternalServerError)
+		return
+	}
+
+	pendingEventsAvailable := (pendingStagingFileCount + pendingUploadCount) > 0
+	triggerPendingUpload, _ := strconv.ParseBool(r.URL.Query().Get(triggerUploadQPName))
+
+	if pendingEventsAvailable && triggerPendingUpload {
+		a.logger.Infof("Triggering upload for all wh destinations connected to source '%s'", sourceID)
+
+		wh := a.bcManager.WarehousesBySourceID(sourceID)
+
+		if len(wh) == 0 {
+			a.logger.Warnf("no warehouse destinations found for source id '%s'", sourceID)
+			http.Error(w, "no warehouse found", http.StatusBadRequest)
+			return
+		}
+
+		for _, warehouse := range wh {
+			triggerUpload(warehouse)
+		}
+	}
+
+	resBody, err := json.Marshal(pendingEventsResponse{
+		PendingEvents:            pendingEventsAvailable,
+		PendingStagingFilesCount: pendingStagingFileCount,
+		PendingUploadCount:       pendingUploadCount,
+		AbortedEvents:            abortedUploadCount > 0,
+	})
+	if err != nil {
+		a.logger.Errorf("unmarshall response for pending events: %v", err)
+		http.Error(w, "can't unmarshall response", http.StatusInternalServerError)
+		return
+	}
+
+	_, _ = w.Write(resBody)
+}
+
+func (a *Api) triggerUploadHandler(w http.ResponseWriter, r *http.Request) {
+	a.logger.LogRequest(r)
+
+	defer func() { _ = r.Body.Close() }()
+
+	var triggerUploadReq triggerUploadRequest
+
+	if err := json.NewDecoder(r.Body).Decode(&triggerUploadReq); err != nil {
+		a.logger.Errorf("unmarshalling body for trigger upload: %v", err)
+		http.Error(w, "can't unmarshall body", http.StatusBadRequest)
+		return
+	}
+
+	sourceID, destinationID := triggerUploadReq.SourceID, triggerUploadReq.DestinationID
+
+	workspaceID, err := a.tenantManager.SourceToWorkspace(r.Context(), sourceID)
+	if err != nil {
+		a.logger.Errorf("checking if source is degraded: %v", err)
+		http.Error(w, "workspaceID from sourceID not found", http.StatusBadRequest)
+		return
+	}
+
+	if a.tenantManager.DegradedWorkspace(workspaceID) {
+		a.logger.Infof("workspace (id: %q) is degrade", workspaceID)
+		http.Error(w, "workspace is degraded", http.StatusServiceUnavailable)
+		return
+	}
+
+	var wh []model.Warehouse
+	if sourceID != "" && destinationID == "" {
+		wh = a.bcManager.WarehousesBySourceID(sourceID)
+	}
+	if destinationID != "" {
```

Comment:
[optional] I think it'd be more readable with a `else if`, just personal preference though.